File: backend/src/models/applications.py
from enum import Enum
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, Field


class ApplicationStatus(str, Enum):
    WISHLIST = "wishlist"
    APPLIED = "applied"
    INTERVIEWING = "interviewing"
    OFFERED = "offered"
    REJECTED = "rejected"
    WITHDRAWN = "withdrawn"


class JobApplication(BaseModel):
    id: Optional[str] = Field(default=None, alias="_id")
    user_id: str
    company: str
    position: str
    job_url: Optional[str] = None
    job_description: Optional[str] = None
    status: ApplicationStatus = ApplicationStatus.WISHLIST
    salary_min: Optional[float] = None
    salary_max: Optional[float] = None
    location: Optional[str] = None
    remote: Optional[bool] = None
    applied_date: Optional[datetime] = None
    notes: Optional[str] = None
    contact_name: Optional[str] = None
    contact_email: Optional[str] = None
    resume_id: Optional[str] = None
    cover_letter_id: Optional[str] = None
    tags: List[str] = []
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        populate_by_name = True
        json_encoders = {datetime: lambda v: v.isoformat()}


class ApplicationCreateRequest(BaseModel):
    company: str
    position: str
    job_url: Optional[str] = None
    job_description: Optional[str] = None
    status: ApplicationStatus = ApplicationStatus.WISHLIST
    salary_min: Optional[float] = None
    salary_max: Optional[float] = None
    location: Optional[str] = None
    remote: Optional[bool] = None
    applied_date: Optional[datetime] = None
    notes: Optional[str] = None
    contact_name: Optional[str] = None
    contact_email: Optional[str] = None
    tags: List[str] = []


class ApplicationUpdateRequest(BaseModel):
    company: Optional[str] = None
    position: Optional[str] = None
    job_url: Optional[str] = None
    job_description: Optional[str] = None
    status: Optional[ApplicationStatus] = None
    salary_min: Optional[float] = None
    salary_max: Optional[float] = None
    location: Optional[str] = None
    remote: Optional[bool] = None
    applied_date: Optional[datetime] = None
    notes: Optional[str] = None
    contact_name: Optional[str] = None
    contact_email: Optional[str] = None
    resume_id: Optional[str] = None
    cover_letter_id: Optional[str] = None
    tags: Optional[List[str]] = None


class ApplicationResponse(BaseModel):
    id: str
    user_id: str
    company: str
    position: str
    job_url: Optional[str] = None
    job_description: Optional[str] = None
    status: ApplicationStatus
    salary_min: Optional[float] = None
    salary_max: Optional[float] = None
    location: Optional[str] = None
    remote: Optional[bool] = None
    applied_date: Optional[datetime] = None
    notes: Optional[str] = None
    contact_name: Optional[str] = None
    contact_email: Optional[str] = None
    resume_id: Optional[str] = None
    cover_letter_id: Optional[str] = None
    tags: List[str] = []
    created_at: datetime
    updated_at: datetime

    class Config:
        populate_by_name = True


class ApplicationStats(BaseModel):
    total: int = 0
    wishlist: int = 0
    applied: int = 0
    interviewing: int = 0
    offered: int = 0
    rejected: int = 0
    withdrawn: int = 0
    response_rate: float = 0.0  # (interviewing + offered) / applied
    offer_rate: float = 0.0     # offered / total

File: frontend/src/services/applications.ts
import api from '../api';

export type ApplicationStatus = 'wishlist' | 'applied' | 'interviewing' | 'offered' | 'rejected' | 'withdrawn';

export interface JobApplication {
  id: string;
  user_id: string;
  company: string;
  position: string;
  job_url?: string;
  job_description?: string;
  status: ApplicationStatus;
  salary_min?: number;
  salary_max?: number;
  location?: string;
  remote?: boolean;
  applied_date?: string;
  notes?: string;
  contact_name?: string;
  contact_email?: string;
  resume_id?: string;
  cover_letter_id?: string;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface ApplicationCreatePayload {
  company: string;
  position: string;
  job_url?: string;
  job_description?: string;
  status?: ApplicationStatus;
  salary_min?: number;
  salary_max?: number;
  location?: string;
  remote?: boolean;
  applied_date?: string;
  notes?: string;
  contact_name?: string;
  contact_email?: string;
  tags?: string[];
}

export interface ApplicationUpdatePayload {
  company?: string;
  position?: string;
  job_url?: string;
  job_description?: string;
  status?: ApplicationStatus;
  salary_min?: number;
  salary_max?: number;
  location?: string;
  remote?: boolean;
  applied_date?: string;
  notes?: string;
  contact_name?: string;
  contact_email?: string;
  resume_id?: string;
  cover_letter_id?: string;
  tags?: string[];
}

export interface ApplicationStats {
  total: number;
  wishlist: number;
  applied: number;
  interviewing: number;
  offered: number;
  rejected: number;
  withdrawn: number;
  response_rate: number;
  offer_rate: number;
}

export const applicationsService = {
  async listApplications(status?: ApplicationStatus, search?: string): Promise<JobApplication[]> {
    const params: Record<string, string> = {};
    if (status) params.status = status;
    if (search) params.search = search;
    const response = await api.get<JobApplication[]>('/applications', { params });
    return response.data;
  },

  async getStats(): Promise<ApplicationStats> {
    const response = await api.get<ApplicationStats>('/applications/stats');
    return response.data;
  },

  async getApplication(id: string): Promise<JobApplication> {
    const response = await api.get<JobApplication>(`/applications/${id}`);
    return response.data;
  },

  async createApplication(payload: ApplicationCreatePayload): Promise<JobApplication> {
    const response = await api.post<JobApplication>('/applications', payload);
    return response.data;
  },

  async updateApplication(id: string, payload: ApplicationUpdatePayload): Promise<JobApplication> {
    const response = await api.patch<JobApplication>(`/applications/${id}`, payload);
    return response.data;
  },

  async deleteApplication(id: string): Promise<void> {
    await api.delete(`/applications/${id}`);
  },
};
